pkg/components/rook/component_test.go

```diff
@@ -0,0 +1,221 @@
+	"github.com/kinvolk/lokomotive/pkg/components/util"
+)
+
+func renderManifest(t *testing.T, configHCL string) map[string]string {
+	name := "rook"
+
+	component, err := components.Get(name)
+	if err != nil {
+		t.Fatalf("could not get the component %q: %v", name, err)
+	}
+
+	body, diagnostics := util.GetComponentBody(configHCL, name)
+	if diagnostics != nil {
+		t.Fatalf("Error getting component body: %v", diagnostics)
+	}
+
+	diagnostics = component.LoadConfig(body, &hcl.EvalContext{})
+	if diagnostics.HasErrors() {
+		t.Fatalf("Valid config should not return error, got: %s", diagnostics)
+	}
+
+	ret, err := component.RenderManifests()
+	if err != nil {
+		t.Fatalf("Rendering manifests with valid config should succeed, got: %s", err)
+	}
+
+	return ret
+}
+
+func getOperatorDeployment(t *testing.T, m map[string]string) *appsv1.Deployment {
+	deployStr, ok := m["rook-ceph/templates/deployment.yaml"]
+	if !ok {
+		t.Fatalf("operator deployment config not found")
+	}
+
+	deploy := &appsv1.Deployment{}
+	if err := yaml.Unmarshal([]byte(deployStr), deploy); err != nil {
+		t.Fatalf("failed unmarshaling manifest: %v", err)
+	}
+
+	return deploy
+}
+
+// nolint:funlen
+func TestConversion(t *testing.T) {
+	configHCL := `
+component "rook" {
+  enable_monitoring = true
+
+  node_selector = {
+    "storage_node_key1" = "storage_node_value1"
+    "storage_node_key2" = "storage_node_value2"
+  }
+
+  toleration {
+    key      = "toleration_key1"
+    operator = "Equal"
+    value    = "toleration_value1"
+    effect   = "NoSchedule"
+  }
+
+  toleration {
+    key      = "toleration_key2"
+    operator = "Equal"
+    value    = "toleration_value2"
+    effect   = "NoSchedule"
+  }
+
+  agent_toleration_key    = "agent_toleration_key"
+  agent_toleration_effect = "NoSchedule"
+
+  discover_toleration_key    = "discover_toleration_key"
+  discover_toleration_effect = "NoSchedule"
+
+  csi_plugin_node_selector = {
+    "node_key1" = "node_value1"
+    "node_key2" = "node_value2"
+  }
+
+  csi_plugin_toleration {
+    key = "other_node_key1"
+    operator = "Equal"
+    value = "other_node_value1"
+    effect = "NoSchedule"
+  }
+}
+`
+
+	m := renderManifest(t, configHCL)
+	if len(m) == 0 {
+		t.Fatalf("Rendered manifests shouldn't be empty")
+	}
+
+	deploy := getOperatorDeployment(t, m)
+
+	tcs := []struct {
+		Name string
+		Test func(*testing.T, map[string]string)
+	}{
+		{"MemoryResources", func(t *testing.T, m map[string]string) {
+			expected := "512Mi"
+			got := deploy.Spec.Template.Spec.Containers[0].Resources.Limits.Memory().String()
+			if expected != got {
+				t.Fatalf("expected: %s, got: %s", expected, got)
+			}
+		}},
+		{"NodeSelector", func(t *testing.T, m map[string]string) {
+			expected := map[string]string{
+				"storage_node_key1": "storage_node_value1",
+				"storage_node_key2": "storage_node_value2",
+			}
+			got := deploy.Spec.Template.Spec.NodeSelector
+
+			if !reflect.DeepEqual(expected, got) {
```

Comment:
I wonder if we can structure the tests to be fully table driven. This means that the `tcs` struct contains *everything* that's specific to a test case: things like initial state, input for the tested function and desired result. You can then call `DeepEqual()` once and you don't have to specify a *function* for each test case. Here is an example of what I mean from MetalLB: https://github.com/metallb/metallb/blob/ec4eee570c96d56a18ccc450e1ede93bf7200876/speaker/bgp_controller_test.go#L209

Reply:
This is true when the output is same. Here we are trying to match different types of the output. Hence I didn't keep a struct field for that.